Add compatibility with PHP 7.3

mickaelandrieu opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Dropped PHP5 support and upgraded the required dependencies
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #12461
| How to test?  | Tests should pass. Please perform a full app test (install + check nothing is broken)


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/13761)
<!-- Reviewable:end -->

mickaelandrieu committed:
Dropped PHP5 support

Update SwiftMailer
Use v4.0.1 of prestashop/translationtools-bundle
Use fromHtml instead of CssToAttributeConverter::__construct

mickaelandrieu commented on composer.json:
```diff
@@ -93,27 +91,32 @@
+        },
+        {
+            "type": "vcs",
+            "url": "https://github.com/mickaelandrieu/TranslationToolsBundle"
```
waiting for https://github.com/PrestaShop/TranslationToolsBundle/pull/57 to be merged

mickaelandrieu replied:
solved! thanks @eternoendless

mickaelandrieu commented on src/PrestaShopBundle/Service/DataProvider/Marketplace/ApiClient.php:
```diff
@@ -64,7 +64,7 @@ public function __construct(
     {
         $this->toolsAdapter->refreshCaCertFile();
-        $this->addonsApiClient->setDefaultOption('verify', $verifySsl);
+        //$this->addonsApiClient->setDefaultOption('verify', $verifySsl);
```
and here it is: in the version of Guzzle required by the Bundle using PHP 7.1, there is no such way to refresh the `verify` option. Is it an issue @Quetzacoalt91?

I may rewrite the client entirely if required, but if it's minor I think we can merge and deprecate ApiClient::setSslVerification().

Quetzacoalt91 replied:
I remember Windows users having issues with SSL check, and this will have to be reviewed with them.
I advise you to remove this line if it does not exist anymore with the new version, but keep the setter in mind in case we have to find another solution.

mickaelandrieu replied:
ok, thanks for the information :)

mickaelandrieu commented on composer.json:
```diff
@@ -93,27 +91,32 @@
-        "symfony/symfony": ">=3.4.8",
+        "symfony/swiftmailer-bundle": "^3.1",
+        "symfony/symfony": "^3.4",
+        "symfony/twig-bridge": "^3.4",
```
not sure we need this

eternoendless replied:
Then why add it?

mickaelandrieu replied:
removed, thanks

kpodemski commented:
what is this? is this will be merged for 1.7.7 or something?

mickaelandrieu commented:
> what is this? is this will be merged for 1.7.7 or something?

:zipper_mouth_face: it's not my place to answer this question but...

mickaelandrieu commented:
Ping @jolelievre, I had to revert the final class on the entity: this is forbidden by Doctrine and on the last updates it throws an exception => https://github.com/doctrine/orm/issues/7598

PierreRambaud commented on composer.json:
```diff
@@ -3,37 +3,35 @@
     "type": "project",
     "require": {
-        "php": ">=5.6.0",
+        "php": ">=7.1.3",
```
```suggestion
        "php": ">=7.1.0",
```

mickaelandrieu replied:
not sure it's worth as most of the libraries set 7.1.3 to be compatible with Symfony 4

PierreRambaud commented on composer.json:
```diff
@@ -93,27 +91,32 @@
         "prestashop/statsstock": "^2",
         "prestashop/statsvisits": "^2",
-        "prestashop/translationtools-bundle": "^3",
+        "prestashop/translationtools-bundle": "dev-sf4-support",
```
https://github.com/PrestaShop/PrestaShop/pull/13761/files#r283138318

PierreRambaud replied:
Because waiting for TranslationToolsBundle pr to be merged

mickaelandrieu replied:
done, solved!

PierreRambaud commented on composer.json:
```diff
@@ -123,7 +126,7 @@
         "mikey179/vfsstream": "^1.6",
         "phake/phake": "@stable",
         "phpunit/phpunit": "~5.7",
```
Update this version too, this phpunit version is really outdated

mickaelandrieu replied:
this is not required, but I'm not against it: should we do it on this pull request too?

PierreRambaud replied:
Yes, we must update all dependencies (as much as we can cf: the mail I forwarded you)

mickaelandrieu replied:
Your mail (if we're talking about the same) don't mention PHPUnit.

Let's discuss it Wednesday with @eternoendless: I have the feeling we're not quite sure of the scope of this PHP update.

Don't get me wrong: I'm also in favor of upgrading PHPUnit to let's say PHPUnit 7 at least (8 requires PHP 7.2), but I'm worried about having to maintain a BIG pull request while my working time is not really dedicated to development tasks 😟.

We can still complete this one with more later 👍

PierreRambaud replied:
phpunit code coverage is not compatible with 7.3 so we need to update it 😎

mickaelandrieu replied:
we don't use phpunit code coverage :trollface:

PierreRambaud replied:
I use it on facetedsearch, and it's already a dependency of phpunit https://packagist.org/packages/phpunit/phpunit#5.7.27 :trollface:

eternoendless replied:
I fear updating phpunit could break existing tests, can we do that on another PR?

mickaelandrieu replied:
I'm also in favor of doing the upgrade of PHPUnit tests in another PR because if "I" do it, I'll use the upgrade of PHPUnit version to review some use of assertions deprecated in the latest versions.

So the upgrade may create a big pull request

PierreRambaud commented on composer.json:
```diff
@@ -203,10 +206,12 @@
         "git-hook-uninstall": "@php .github/contrib/uninstall.php"
     },
+    "scripts-descriptions": {
+      "test-all": "Launch all PHPUnit test suites",
```
Why two spaces here when the whole file is four :D

mickaelandrieu replied:
it's my PHPStorm setup :p, I'll revert that: let's see how @eternoendless will react to this pull request first ;)

PierreRambaud replied:
As I said in a mail few days ago, we need to communicate before merging this kind of work 😺

mickaelandrieu replied:
solved!

PierreRambaud commented on composer.json:
```diff
@@ -93,27 +91,32 @@
+        "swiftmailer/swiftmailer": "~6.2",
         "symfony/monolog-bundle": "^3.1.0",
         "symfony/polyfill-apcu": "^1.0",
         "symfony/polyfill-php73": "^1.10",
```
Are we really using this component? (same for acpu polyfill?)

mickaelandrieu replied:
We probably don't, but how can we be sure now it's here?

PierreRambaud replied:
There are 5 functions / Class this polyfill is taking in consideration, we can maybe try to find if it used 🤔

Quetzacoalt91 replied:
We are, yes. Look at the PR which introduced it. :)

https://github.com/PrestaShop/PrestaShop/pull/11510

PierreRambaud commented on src/PrestaShopBundle/Entity/Lang.php:
```diff
@@ -34,7 +34,7 @@
  * @ORM\Entity(repositoryClass="PrestaShopBundle\Entity\Repository\LangRepository")
  */
-final class Lang implements LanguageInterface
+class Lang implements LanguageInterface
```
Why not final?

mickaelandrieu replied:
See my last comment https://github.com/PrestaShop/PrestaShop/pull/13761#issuecomment-491600332 ;)

mickaelandrieu commented:
ping @eternoendless 

```
Error: Call to undefined method PrestaShop\TranslationToolsBundle\Translation\Helper\DomainHelper::buildModuleBaseDomain()
```

Is it intended? What do I need to do to use the latest version of the Bundle?

eternoendless commented on .travis.yml:
```diff
@@ -75,11 +76,10 @@ before_install:
+    composer update --ignore-platform-reqs --no-suggest --ansi --no-interaction --no-progress --quiet;
     else
-        composer install --no-suggest --ansi --prefer-dist --no-interaction --no-progress --quiet;
+    composer install --no-suggest --ansi --prefer-dist --no-interaction --no-progress --quiet;
```
We can keep the indentation here

mickaelandrieu replied:
indeed, i'll revert that

mickaelandrieu replied:
reverted, solved.

eternoendless commented on app/config/config.yml:
```diff
@@ -117,9 +117,6 @@ swiftmailer:
         enabled: "%kernel.debug%"
-    cache:
-        enabled: true
-        adapter: guzzle.cache
```
Is this change related?

mickaelandrieu replied:
yes, the new version of the bundle doesn't accept that configuration.

I haven't investigated so far on this, mostly because I think we will use the circuit breaker to query the list of modules and use the integrated cache of Symfony instead: am I wrong?

eternoendless commented on composer.json:
```diff
@@ -3,37 +3,35 @@
         "geoip2/geoip2": "~2.4.2",
         "greenlion/php-sql-parser": "^4.1",
-        "guzzlehttp/guzzle": "~5.0",
+        "guzzlehttp/guzzle": "~6.1",
```
Can't we keep 5.x? It seems that many modules use Guzzle 5 to be compatible with previous PS versions

kpodemski replied:
about that, is it possible in 1.7.6 to use our own Guzzle version declared in composer, in module? i guess... not? i remember few discussions on StackOverflow, when developer needed to use different version of Guzzle in his own module but it was impossible due to conflicts

mickaelandrieu replied:
> Can't we keep 5.x? It seems that many modules use Guzzle 5 to be compatible with previous PS versions

Sadly, I don't think it's possible, except if we remove the guzzle bundle and create our own services "hand by hand". See the requirements of the bundle on packagist => https://packagist.org/packages/csa/guzzle-bundle

How many modules are involved? Can't we upgrade them too?

@kpodemski I don't know, but I don't see a valid reason to use a specific version of guzzle when the core already provide one :/

eternoendless replied:
Upgrading Guzzle will be a major problem for modules who need to target all our current installed base, the will need to distribute separate zip files for PS <= 1.7.6

Is there a specific reason to update guzzle-bundle? Because otherwise we'll have to fork it. I think upgrading Guzzle in 1.7 is not a good idea.

mickaelandrieu replied:
> Is there a specific reason to update guzzle-bundle

yes, the minimal php version supported :+1: 

> I think upgrading Guzzle in 1.7 is not a good idea.

I'm not fan of that too but in all cases people will complain:

* the ones who use v6 and are stuck with v5 in the core;
* the ones who use v5 and who will cry if we update;

This is why I asked for existing modules using Guzzle, because the upgrading guide shows that most of the public APIs are kept between v5 and v6. The impacts are null for people using it as a PHP REST client => https://github.com/guzzle/guzzle/blob/master/UPGRADING.md#50-to-60

> Because otherwise, we'll have to fork it.

I'll advise to not rely on a bundle anymore if we're stuck with an older version that we can't upgrade

eternoendless commented on composer.json:
```diff
@@ -93,27 +91,32 @@
         "smarty/smarty": "^3.1.31",
         "soundasleep/html2text": "^0.5.0",
-        "swiftmailer/swiftmailer": "~5.4.3",
+        "swiftmailer/swiftmailer": "~6.2",
```
Do we need to update this package?

mickaelandrieu replied:
yes, but not because of PHP requirements but because this major release fixes bug we had in the core and that we couldn't fix. I can't remember the related issues, but @PierreRambaud probably do? Also, as you can see, this upgrade has no impact at all on our public API :)

PierreRambaud replied:
It does not support recent versions of the email adress RFC, have many bugs with punycode and validators.
But v5.4.3 requires php: >=5.3.3 where 6.2.1 requires php: >=7.0.0, I think we could have some issues with 7.3 if we don't update

eternoendless replied:
👍

eternoendless commented on composer.json:
```diff
@@ -147,10 +150,10 @@
+          "PrestaShopBundle\\Install\\Upgrade::migrateSettingsFile",
+          "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
+          "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
+          "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
```
Wrong indentation here

mickaelandrieu replied:
I'll revert that, don't worry

mickaelandrieu replied:
done

eternoendless commented on install-dev/controllers/console/process.php:
```diff
@@ -350,7 +350,6 @@ private function initKernel()
         }
         global $kernel;
         $kernel = new AppKernel($env, _PS_MODE_DEV_);
-        $kernel->loadClassCache();
```
Why delete this line?

mickaelandrieu replied:
This was an optimization not useful when using PHP 7 or above => https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/HttpKernel/Kernel.php#L394

eternoendless commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/MailTheme/Blocks/configuration_form.html.twig:
```diff
@@ -22,6 +22,7 @@
  * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
  * International Registered Trademark & Property of PrestaShop SA
  *#}
+{% import '@PrestaShop/Admin/macros.html.twig' as ps %}
```
Is this for making an implicit import explicit?

mickaelandrieu replied:
This is required starting from Twig 2, as macro functions are not shared across children templates anymore. According to the @twigphp/Twig core team, this was a side effect of Twig 1.x and never intended to be used this way.

Sadly [we made the related test suite optional](https://github.com/PrestaShop/PrestaShop/blob/develop/.travis.yml#L64) so I had to fix it in this pull request when before we never had to think about it: technical debt have to be paid.

Better now than the day of no one can figure out why it's broken :sunglasses:

mickaelandrieu commented:
Thanks for the review @eternoendless.

I tried to answer your questions the best I could, let me know what I can do to make this pull request merged ASAP.

But before that, we need the pull request on the TranslationToolsBundle to be accepted :+1: 

Last comment:

> Then why add it?

This is a memory of my previous attempts to migrate to Symfony 4. As it's more difficult than I've hoped to be, and as it's not in the scope of 1.7 I'll revert that too.

caligueco commented:
Hello. I need help with my store prestashop 1.6. Yesterday I updated the php version from the cpanel of my hosting from 5.4.45 to 5.6 and from that moment I can not enter the back office to manage the store. I'm desperate. how can I solve that?. Can I edit a prestashop file to fix it? Can I go back somehow to the php version 5.4? the system does not allow it, it only lets me update to 5.5, 7.0, 7.1, 7.2 and 7.3.
a greeting

matks commented:
> Hello. I need help with my store prestashop 1.6. Yesterday I updated the php version from the cpanel of my hosting from 5.4.45 to 5.6 and from that moment I can not enter the back office to manage the store. I'm desperate. how can I solve that?. Can I edit a prestashop file to fix it? Can I go back somehow to the php version 5.4? the system does not allow it, it only lets me update to 5.5, 7.0, 7.1, 7.2 and 7.3.
> a greeting

Hi, I'm sorry but this is an issue you can only solve with your hosting provider, prestashop has nothing to do with how your hosting provider manages php versions. Please be aware that php5 versions are now EOL and not maintained anymore, you should upgrade to php7 as soon as possible and upgrade your PS store to the latest PS1.6 version which allows this usage. You need to reach out to your hosting provider and discuss with him how you can do that.

mickaelandrieu commented on src/PrestaShopBundle/Service/DataProvider/Marketplace/ApiClient.php:
```diff
@@ -64,6 +64,7 @@ public function __construct(
     public function setSslVerification($verifySsl)
     {
         $this->toolsAdapter->refreshCaCertFile();
+        $this->addonsApiClient->setDefaultOption('verify', $verifySsl);
```
I can introduce it back now we're using Guzzle 5 again :)

mickaelandrieu commented:
Long story short: we're **forking** CsaGuzzleBundle to allow PHP 7.1+ support with Guzzle 5.

See my last commit for more information.

mickaelandrieu commented:
one more issue to fix, coming from TranslationToolsBundle 👍

matks commented:
Build fails because of bad catalog extraction, should be fixed in https://github.com/PrestaShop/TranslationToolsBundle/pull/61

matks commented on app/config/config.yml:
```diff
@@ -114,12 +114,12 @@ swiftmailer:
-csa_guzzle:
-    profiler:
-        enabled: "%kernel.debug%"
+love_oss_guzzle5:
```
Seen with @eternoendless and @PierreRambaud : this is a BC break as it will fail for people using this configuration key.

Example in this PR: https://github.com/PrestaShop/PrestaShop/pull/13761/files#diff-95a1b476935b61587950fe999d326bbaR165
If someone was using `$clientConfig = $config['csa_guzzle']['clients']['addons_api']['config'];` it will fail after this PR

matks commented on app/config/config.yml:
```diff
@@ -114,12 +114,12 @@ swiftmailer:
-csa_guzzle:
-    profiler:
-        enabled: "%kernel.debug%"
+love_oss_guzzle5:
```
Seen with @eternoendless and @PierreRambaud : this is a BC break as it will fail for people using this configuration key.

Example in this PR: https://github.com/PrestaShop/PrestaShop/pull/13761/files#diff-95a1b476935b61587950fe999d326bbaR165
If someone was using `$clientConfig = $config['csa_guzzle']['clients']['addons_api']['config'];` it will fail after this PR

matks replied:
@mickaelandrieu Unfortunately because of this issue, we'll have to re-create a fork of CsaGuzzleBundle but this time we'll try to make it as close as possible to the original project: no namespace change, no config key change. The aim is really to keep the code updated and backward compatible

In order to do this, a PrestaShop fork is on the way : https://github.com/PrestaShop/CsaGuzzleBundle

mickaelandrieu replied:
I'm not sure we have the right to do that (use the author's name) and this will confuse people who will try go get support to the original github repository: the code is open source, but Cza's name is probably not ? /c @ttoine 

> If someone was using $clientConfig = $config['csa_guzzle']['clients']['addons_api']['config']; it will fail after this PR

We can probably alias it for backward compatibility. I'm not in favor of a fake fork but in the end, it's your choice: I'm not part of the maintainers anymore 👼

ttoine replied:
a nickname is usually considered as public.
however, I don't understand why/if this is useful to use/quote this nickname in code ? maybe it is better to write a clear reference to the origin of this config in the comment; like "# Guzzle configuration inspired by csa's"?

matks replied:
🆙 @ttoine do you have any advice on this authoring matter ? 🤔

ttoine replied:
please see my review.

ttoine reviewed (dismissed):
Todo:
- original library is in Apache v2 (I can't understand why, for both Symfony and Guzzle are in MIT), but please put the right license for the fork, this Apache v2 too; you can't change the license this way.
- change the name of the library in the headers, readme, etc. but keep the original name where needed in the code for retro-compatibility. 
- add a paragraph in the main readme.md to explain that the name is kept for the transition for retro-compatibility purpose.

mickaelandrieu commented:
Thanks for you review @ttoine, as @matks plan to use PrestaShop's fork, he may address all your comments to do it the right way :+1: .

Once it's done, I'll delete my fork: I have no interest in maintenance of a bundle for a deprecated version of Guzzle.

ping @eternoendless: once 1.7.6.0 release is done, we should finish and merge this PR ASAP! I may have broken some stuff. The sooner this breaks, the better for everyone :angel:

matks commented:
https://github.com/PrestaShop/CsaGuzzleBundle should be now a fork of CsaGuzzleBundle compatible with php7.3 thanks to @Progi1984 

We can now test this PR targeting this repo.

matks commented:
@ttoine the current license header for https://github.com/PrestaShop/CsaGuzzleBundle is
```
/*
 * This file is part of the CsaGuzzleBundle package
 *
 * (c) Charles Sarrazin <charles@sarraz.in>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code
 */
```
If I change it to
```
/*
 * This file is part of the CsaGuzzleBundle package, forked by PrestaShop
 *
 * (c) Charles Sarrazin <charles@sarraz.in>
 *  PrestaShop SA <contact@prestashop.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code
 */
```
Is that correct ?

matks commented:
Looks like Travis lost this build 😒

PierreRambaud commented:
@matks You're working too much, travis banned you? :trollface:

matks commented on composer.lock:
```diff
@@ -5451,29 +5844,37 @@
         {
             "name": "swiftmailer/swiftmailer",
-            "version": "v5.4.12",
+            "version": "v6.2.1",
```
Might introduce breaking changes 🤔

matks commented on composer.lock:
```diff
@@ -6024,44 +6779,50 @@
         {
             "name": "symfony/swiftmailer-bundle",
-            "version": "v2.6.7",
+            "version": "v3.2.8",
```
Breaking change possible

matks commented on composer.lock:
```diff
@@ -6347,31 +7108,32 @@
         {
             "name": "twig/twig",
-            "version": "v1.38.4",
+            "version": "v2.11.3",
```
Breaking change possible

matks commented on composer.lock:
```diff
@@ -6024,44 +6645,47 @@
         {
             "name": "symfony/swiftmailer-bundle",
-            "version": "v2.6.7",
+            "version": "v3.2.6",
```
Possible breaking change

matks commented on composer.lock:
```diff
@@ -6347,21 +6971,22 @@
         {
             "name": "twig/twig",
-            "version": "v1.38.4",
+            "version": "v2.10.0",
```
Possible breaking change

matks commented:
Summary of risky composer dependencies updates:

#### Our repositories
- "csa/guzzle-bundle": 1.3.12 -> 1.3.13, using our own fork
- "prestashop/translationtools-bundle": 3.1.0 -> 4.0.1
#### Doctrine
- "doctrine/annotations": 1.4.0 -> 1.6.1
- "doctrine/cache": 1.6.2 -> 1.8.0
- "doctrine/collections": 1.4.0 -> 1.6.1
- "doctrine/common": 2.5.3 -> 2.10.0
- "doctrine/dbal": 2.5.13 -> 2.9.2
- "doctrine/doctrine-bundle": 1.10.2 -> 1.11.1
- "doctrine/inflector":  1.1.0 -> 1.3.0
- "doctrine/instantiator": 1.0.5 -> 1.2.0
- "doctrine/orm": 2.5.14 -> 2.6.3
#### Other
- "beberlei/doctrineextensions": 1.1.9 -> 1.2.1
- "league/tactician-bundle": 0.4.1 -> 1.1.16 ⚠️ 
- "nikic/php-parser": 3.1.5 -> 4.2.1 ⚠️ 
- "sensio/framework-extra-bundle": 5.2.4 -> 5.3.1
- "swiftmailer/swiftmailer": 5.4.12 -> 6.2.1 ⚠️ 
- "symfony/swiftmailer-bundle": 2.6.7 -> 3.2.6 ⚠️ 
- "symfony/symfony": 3.4.26 -> 3.4.27
- "twig/twig": 1.38.4 -> 2.10.0 ⚠️ 

@PierreRambaud @eternoendless

matks commented:
> Everything must be updated if possible.
> For example, with the latest swiftmailer, it uses the latest email RFC.
> And we must do the same thing for phpunit (which is a very old version :/)

Isn't it going to introduce breaking changes ? For example twig v1 -> twig v2 ?

ebatarson-axome commented:
Hi,
Is it possible to upgrade this dependencies ?
- Guzzle 6+ 
- Doctrine : doctrine-common >= 2.6.x (issue with type hint / Proxy class in < 2.6

PierreRambaud commented:
> > Everything must be updated if possible.
> > For example, with the latest swiftmailer, it uses the latest email RFC.
> > And we must do the same thing for phpunit (which is a very old version :/)
> 
> Isn't it going to introduce breaking changes ? For example twig v1 -> twig v2 ?

Updating dependencies without breaking our public API doesn't mean it's a BC break.
If we communicate right now what we are going to change, especially for twig, I think i's ok. More because upgrading from twig 1 to twig 2 is really easy.

matks commented:
> > > Everything must be updated if possible.
> > > For example, with the latest swiftmailer, it uses the latest email RFC.
> > > And we must do the same thing for phpunit (which is a very old version :/)
> > 
> > 
> > Isn't it going to introduce breaking changes ? For example twig v1 -> twig v2 ?
> 
> Updating dependencies without breaking our public API doesn't mean it's a BC break.
> If we communicate right now what we are going to change, especially for twig, I think i's ok. More because upgrading from twig 1 to twig 2 is really easy.

Then why did we fork CsaGuzzleBundle to maintain a 1.3 version compatible with php7.3 instead of upgrading it 😅?

It seems we disagree on what is a Breaking Change. @eternoendless we need your feedback here

PierreRambaud commented:
> > > > Everything must be updated if possible.
> > > > For example, with the latest swiftmailer, it uses the latest email RFC.
> > > > And we must do the same thing for phpunit (which is a very old version :/)
> > > 
> > > 
> > > Isn't it going to introduce breaking changes ? For example twig v1 -> twig v2 ?
> > 
> > 
> > Updating dependencies without breaking our public API doesn't mean it's a BC break.
> > If we communicate right now what we are going to change, especially for twig, I think i's ok. More because upgrading from twig 1 to twig 2 is really easy.
> 
> Then why did we fork CsaGuzzleBundle to maintain a 1.3 version compatible with php7.3 instead of upgrading it ?
> 
> It seems we disagree on what is a Breaking Change. @eternoendless we need your feedback here

Unlike Twig, CsaGuzzle totally breaks public API and a huge refactoring is needed to make it compliant with the latest version. With twig, your version 1 templates are also valid with twig 2 (only few things like global macros need changes)

matks commented:
We need to update composer.lock PrestaShop/CsaGuzzleBundle target to https://github.com/PrestaShop/CsaGuzzleBundle/commit/df7d5a9f3c37be05463cb9a54c186827aba19862 and then we're good for CsaGuzzleBundle 👍

eternoendless commented:
We should avoid upgrading libraries that include breaking changes, unless it's absolutely needed to ensure compatibility with PHP 7.3, **specially** if those libraries are usually used by modules.

matks commented:
OK

- Doctrine dependencies upgrade is required
- I think beberlei/doctrineextensions and nikic/php-parser are needed by Doctrine and TranslationToolsBundle

So this we can keep as it is

The followings, I think they do not introduce breaking changes, but we need to be sure:
- league/tactician-bundle
- sensio/framework-extra-bundle
- symfony/symfony
So after we check, we can keep them

The followings, I think they introduce breaking changes, but we need to check:
- swiftmailer/swiftmailer
- symfony/swiftmailer-bundle
- twig/twig

PierreRambaud commented:
swiftmailer/swiftmailer is required. It fixed many bugs and issues (because we have a really really old version which following an old RFC + punycode), and our version is not mark as compatible with 7.3.

Plus, we must upgrade all dev-dependencies. 
For example, phpunit have many functions broken in 7.3 :/

matks commented:
> Plus, we must upgrade all dev-dependencies.

I think SemVer only applies to the zip we build, for dev dependencies we can do whatever we need.

mickaelandrieu commented:
> "twig/twig": 1.38.4 -> 2.10.0 :warning:

it's not totally true :) [before](https://github.com/PrestaShop/PrestaShop/blob/develop/composer.json#L112) it was "1.38||2.6" and now it is "2.10": we support officially Twig 2 since PrestaShop 1.7.3.

matks commented:
> > "twig/twig": 1.38.4 -> 2.10.0 ⚠️
> 
> it's not totally true :) [before](https://github.com/PrestaShop/PrestaShop/blob/develop/composer.json#L112) it was "1.38||2.6" and now it is "2.10": we support officially Twig 2 since PrestaShop 1.7.3, this is also why I've complained so hard when you put the related build optional...

The composer.json is one thing, the composer.lock is another one 😉 
All QA tests are done using the composer.lock version which was Twig 1.38
So by switching to Twig 2.10, every check done by QA is not valid anymore for twig-related pages

matks commented:
For https://github.com/thephpleague/tactician-bundle/ the [Changelog](https://github.com/thephpleague/tactician-bundle/blob/master/CHANGELOG.md) does warn about BC breaks

mickaelandrieu commented:
> All QA tests are done using the composer.lock version which was Twig 1.38

> this is also why I've complained so hard when you put the related build optional... which is using Twig 2

eternoendless commented:
Is it _necessary_ to drop Twig 1 to support PHP 7.3?

If not, then we don't need to do it right now.

matks commented:
> Is it _necessary_ to drop Twig 1 to support PHP 7.3?
> 
> If not, then we don't need to do it right now.

No, it is not necessary.

mickaelandrieu commented:
> No, it is not necessary.

Agree 👍, but this is not "that" easy.

Now we're using PHP 7.1 as minimal, our composer.lock file will get Twig 2 by default: this means we need a build specific to test our project with Twig 1.

PierreRambaud commented:
> > No, it is not necessary.
> 
> Agree , but this is not "that" easy.
> 
> Now we're using PHP 7.1 as minimal, our composer.lock file will get Twig 2 by default: this means we need a build specific to test our project with Twig 1.

We use the composer.lock to create a release, not the composer.json. 
We can update the composer.json and remove the Twig2 check, 1) we don't use it, 2) we don't release it 3) prestashop/prestashop is not on packagist.

mickaelandrieu commented:
> We use the composer.lock to create a release, not the composer.json.

"It's your problem" will say agencies like OhmyWeb (/c @SebBareyre ) that use directly GitHub to deploy PrestaShop. We officially support Twig 1 & 2: if you remove the support of Twig 2 or Twig 1 this needs to be part of the changelog. It's not a BC so it could be done in a minor version.

matks commented:
> We officially support Twig 1 & 2

I'm sorry, no, we unofficially support Twig 2. Not officially, and if one of our statement says the opposite, we need to fix it IMO.

We are trying our very best to make sure people can use upper dependencies for PrestaShop:
- even if we target php 7.1 to php 7.3, if we can start making php7.4-compliant, we'll do it
- even if we target mysql 5.7, if we can make our SQL mysql8.0-compliant, we'll do it
- even if we target Twig 1, if we can make our templates Twig v2-compliant we'll do it
- (I guess) even if we target a Linux-like environment, if we can make our code Windows-compliant, we'll do it
- right now we are trying to enable the use of higher jQuery versions to follow this

By "we'll do it" I mean: "provided this does not make the code a lot more complex and unreadable".

But this is not an official support. This is a friendly and constructive support to help people use higher dependencies.

This is my opinion, I can be wrong, please correct me in this case @eternoendless @Quetzacoalt91 @PierreRambaud . 🤔

mickaelandrieu commented:
> I'm sorry, no, we unofficially support Twig 2

The Composer configuration says we do [since 1.7.4 at least](https://github.com/PrestaShop/PrestaShop/blob/1.7.4.x/composer.json#L105), we can debate a long time about it but if we remove Twig 2 support we must add a note on the changelog.

>  if we can make our templates Twig v2-compliant we'll do it

Our templates were v2-compliants until the core team decides to put a required build optional in 1.7.6. As far as I remember, it was not to remove Twig 2 support but because of false positives tests: I can't remember any discussions about the removal of Twig 2 but I can be wrong here, tell me.

> This is a friendly and constructive support to help people use higher dependencies.

Twig is a not in the same situation of the others dependencies you're listing because we clearly describe in Composer configuration the support of the 2 versions: I'm not sure I'm well understood here so we may take a look at [how versions and constraints are described in Composer](https://getcomposer.org/doc/articles/versions.md#version-range) and discuss about it.

In all cases, if our real "support" policy doesn't fit what is described in the Composer configuration, we need to say it somewhere: that's why I mention an update of the changelog.

matks commented:
Composer ranges are not an official support. An official support is a written statement in the developer documentation.

Composer ranges are a settings used to configure Composer so it knows which dependencies it can pull. It can be used as a version compatibility indicator when the repository is being distributed by packagist. But only as an indicator, not an official support statement.

Example: on branch `1.7.6.x`, our Composer range version says `"php": ">=5.6.0"`,
so according to you it means PS 1.7.6 supports php7.3 😛?

Finally, if we declare "We support environment XYZ" then we should truly do it, meaning we should test our PRs with it, we should ask QA to test PRs with this environment too and, of course, we should test our releases with it. So if we decide to support Twig v2 for 1.7.7, then we need to make sure QA team does validate the 1.7.7 build with Twig v2, not Twig v1.

Without this extended check, saying we support it is meaningless. Either we support something and we actively do it or we dont support something. Being in the middle is wasting both our time and our users' time.

And a Travis build is not enough to validate all our templates are v2 compliant. It's a nice item, but it's not enough. QA full validation is needed and it was never performed for previous PS versions.

mickaelandrieu commented:
> Composer ranges are not an official support

We agree to disagree on that :) note that it's possible to set a range on PHP versions too.

> So if we decide to support Twig v2 for 1.7.7, then we need to make sure QA team does validate the 1.7.7 build with Twig v2, not Twig v1.

Again: a build were done for that already, that the core team have made optional during 1.7.6. I have the feeling you're find a soft way to stop the support of Twig 2.

You don't need to: just update Composer, delete the build AND (really important to avoid shitstorm) add a note on the changelog.

At the end of the day, you own and maintain the project, not me! 

I'm not arguing in favor or against Twig 2 here, I'm arguing _only_ about the fact you want to remove it without noticing people/shops that relies on it since 1.7.4.

I guess I've explained enough the risks and shareholders so I won't argue about it anymore.

Thanks for listening and patience!

PierreRambaud commented on composer.json:
```diff
@@ -123,7 +125,7 @@
         "mikey179/vfsstream": "^1.6",
         "phake/phake": "@stable",
         "phpunit/phpunit": "~5.7",
```
Still not compatible with php 7.3

Commit:
Update PHPUnit configuration

All tests must be compatible with latest php version.
Travis must be focused on the latest release for sanity and
integration tests.

mickaelandrieu commented on composer.json:
```diff
@@ -107,7 +107,7 @@
         "tecnickcom/tcpdf": "^6.2.12",
         "tijsverkoyen/css-to-inline-styles": "^2.2",
-        "twig/twig": "^2.6.2"
+        "twig/twig": "^1.38"
```
this commit name is misleading: you have removed [twig 2 support](https://github.com/PrestaShop/PrestaShop/blob/develop/composer.json#L113).

How about "Removed Twig 2 support" instead?

PierreRambaud replied:
You're still able to change the composer.json if you want to use Twig 2. But we never say we support twig 2, and never release a version with Twig 2.

mickaelandrieu replied:
It's not for me, its for agencies like OhmyWeb for instance: say what you want, I just try to avoid another community shitstorm.

PierreRambaud replied:
This is exactly what I said, if they play with the composer.json / composer.lock, they are able to change dependencies for their usage. But right now, we never said we support Twig2 and we never release Twig2 in the zip file.

PierreRambaud commented:
Required #15902 to continue

matks commented:
Just came back from PHP Forum event, saw this tool to help tracking issues for php versions upgrade: https://github.com/PHPCompatibility/PHPCompatibility

PierreRambaud commented:
@matks Already ran it weeks ago 😉

PierreRambaud commented:
FYI Error with phpcompatibility check.


```
FILE: /app/src/Core/Util/String/StringModifier.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 27 | ERROR | 'string' is a reserved keyword as of PHP version 7.0 and should not be used to name a
    |       | class, interface or trait or as part of a namespace (T_NAMESPACE)
--------------------------------------------------------------------------------------------------------


FILE: /app/src/Core/Util/String/StringValidatorInterface.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 27 | ERROR | 'string' is a reserved keyword as of PHP version 7.0 and should not be used to name a
    |       | class, interface or trait or as part of a namespace (T_NAMESPACE)
--------------------------------------------------------------------------------------------------------


FILE: /app/src/Core/Util/String/StringModifierInterface.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 27 | ERROR | 'string' is a reserved keyword as of PHP version 7.0 and should not be used to name a
    |       | class, interface or trait or as part of a namespace (T_NAMESPACE)
--------------------------------------------------------------------------------------------------------


FILE: /app/src/Core/Util/String/StringValidator.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 27 | ERROR | 'string' is a reserved keyword as of PHP version 7.0 and should not be used to name a
    |       | class, interface or trait or as part of a namespace (T_NAMESPACE)
--------------------------------------------------------------------------------------------------------


FILE: /app/src/Core/String/CharacterCleaner.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 27 | ERROR | 'string' is a reserved keyword as of PHP version 7.0 and should not be used to name a
    |       | class, interface or trait or as part of a namespace (T_NAMESPACE)
--------------------------------------------------------------------------------------------------------


FILE: /app/src/Adapter/Presenter/Object/ObjectPresenter.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 27 | ERROR | 'object' is a soft reserved keyword as of PHP version 7.0 and a reserved keyword as of
    |       | PHP version 7.2 and should not be used to name a class, interface or trait or as part of
    |       | a namespace (T_NAMESPACE)
--------------------------------------------------------------------------------------------------------

FILE: /app/controllers/admin/AdminImportController.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 2634 | ERROR | Extension 'mysql_' is deprecated since PHP 5.5 and removed since PHP 7.0; Use mysqli
      |       | instead
--------------------------------------------------------------------------------------------------------

FILE: /app/classes/Cookie.php
------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
------------------------------------------------------------------------------------
 361 | WARNING | INI directive 'mbstring.func_overload' is deprecated since PHP 7.2
------------------------------------------------------------------------------------


FILE: /app/classes/Tools.php
--------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 4200 | WARNING | Since PHP 7.0, functions inspecting arguments, like func_get_args(), no longer report
      |         | the original value as passed to a parameter, but will instead provide the current
      |         | value. The parameter "$replacements" was used, and possibly changed (by reference),
      |         | on line 4197.
--------------------------------------------------------------------------------------------------------


FILE: /app/classes/PhpEncryptionLegacyEngine.php
--------------------------------------------------------------------------------------------------------
FOUND 13 ERRORS AFFECTING 7 LINES
--------------------------------------------------------------------------------------------------------
  41 | ERROR | The constant "MCRYPT_MODE_CBC" is deprecated since PHP 7.1 and removed since PHP 7.2
  42 | ERROR | The constant "MCRYPT_RIJNDAEL_128" is deprecated since PHP 7.1 and removed since PHP 7.2
  53 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use openssl
     |       | (preferred) or pecl/mcrypt once available instead
  53 | ERROR | Function mcrypt_get_iv_size() is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | OpenSSL instead
  54 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use openssl
     |       | (preferred) or pecl/mcrypt once available instead
  54 | ERROR | Function mcrypt_create_iv() is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | random_bytes() or OpenSSL instead
  54 | ERROR | The constant "MCRYPT_RAND" is deprecated since PHP 7.1 and removed since PHP 7.2
  67 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use openssl
     |       | (preferred) or pecl/mcrypt once available instead
  67 | ERROR | Function mcrypt_get_block_size() is deprecated since PHP 7.1 and removed since PHP 7.2;
     |       | Use OpenSSL instead
  70 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use openssl
     |       | (preferred) or pecl/mcrypt once available instead
  70 | ERROR | Function mcrypt_encrypt() is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | OpenSSL instead
 109 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use openssl
     |       | (preferred) or pecl/mcrypt once available instead
 109 | ERROR | Function mcrypt_decrypt() is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | OpenSSL instead
--------------------------------------------------------------------------------------------------------


FILE: /app/classes/Mail.php
-----------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------------
 902 | WARNING | The constant "INTL_IDNA_VARIANT_2003" is deprecated since PHP 7.2
-----------------------------------------------------------------------------------


FILE: /app/classes/PhpEncryptionEngine.php
--------------------------------------------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 142 | ERROR | Extension 'mcrypt' is deprecated since PHP 7.1 and removed since PHP 7.2; Use openssl
     |       | (preferred) or pecl/mcrypt once available instead
 142 | ERROR | Function mcrypt_create_iv() is deprecated since PHP 7.1 and removed since PHP 7.2; Use
     |       | random_bytes() or OpenSSL instead
 142 | ERROR | The constant "MCRYPT_DEV_URANDOM" is deprecated since PHP 7.1 and removed since PHP 7.2
--------------------------------------------------------------------------------------------------------

```

PierreRambaud commented:
I forget to ran it on our modules :scream: 
But there is nothing specific to php7.3 :)
```
FILE: /app/modules/ps_wirepayment/validation.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 48 | ERROR | Using 'break' outside of a loop or switch structure is invalid and will throw a fatal
    |       | error since PHP 7.0
--------------------------------------------------------------------------------------------------------


FILE: /app/modules/ps_wirepayment/controllers/front/validation.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 47 | ERROR | Using 'break' outside of a loop or switch structure is invalid and will throw a fatal
    |       | error since PHP 7.0
--------------------------------------------------------------------------------------------------------


FILE: /app/modules/ps_wirepayment/controllers/front/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_wirepayment/controllers/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_wirepayment/views/templates/hook/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_wirepayment/views/templates/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_wirepayment/views/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/gamification/classes/GamificationTools.php
--------------------------------------------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 3 LINES
--------------------------------------------------------------------------------------------------------
 45 | ERROR | Indirect access to variables, properties and methods will be evaluated strictly in
    |       | left-to-right order since PHP 7.0. Use curly braces to remove ambiguity.
 48 | ERROR | Indirect access to variables, properties and methods will be evaluated strictly in
    |       | left-to-right order since PHP 7.0. Use curly braces to remove ambiguity.
 51 | ERROR | Indirect access to variables, properties and methods will be evaluated strictly in
    |       | left-to-right order since PHP 7.0. Use curly braces to remove ambiguity.
--------------------------------------------------------------------------------------------------------


FILE: /app/modules/ps_faviconnotificationbo/ps_faviconnotificationbo.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 50 | ERROR | The behaviour of hexadecimal numeric strings was inconsistent prior to PHP 7 and support
    |       | has been removed in PHP 7. Found: '0x64aa3c1e4034d07015f639b0e171b0d7b27d01aa'
--------------------------------------------------------------------------------------------------------


FILE: /app/modules/ps_socialfollow/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_customtext/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_contactinfo/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_categorytree/img/icon/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_categorytree/img/index.php
-----------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
-----------------------------------------------------------------------------
 1 | WARNING | File has mixed line endings; this may cause incorrect results
-----------------------------------------------------------------------------


FILE: /app/modules/ps_themecusto/ps_themecusto.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 51 | ERROR | The behaviour of hexadecimal numeric strings was inconsistent prior to PHP 7 and support
    |       | has been removed in PHP 7. Found: '0x64aa3c1e4034d07015f639b0e171b0d7b27d01aa'
--------------------------------------------------------------------------------------------------------


FILE: /app/modules/blockreassurance/blockreassurance.php
--------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------------------------------
 106 | ERROR | The behaviour of hexadecimal numeric strings was inconsistent prior to PHP 7 and support
     |       | has been removed in PHP 7. Found: '0x64aa3c1e4034d07015f639b0e171b0d7b27d01aa'
--------------------------------------------------------------------------------------------------------

Time: 11.36 secs; Memory: 20MB
```

Commit:
Remove unwanted library because already defined in composer.json plus update parser_sql to latest release to be compatible with php 7.3

Gamesh commented:
>    > so according to you it means PS 1.7.6 supports php7.3 😛?
> 
> Isn't what you'd expect when you use composer install command without any warnings while you're on PHP 7.3 ? 😇

yes, but my two cents here would be to constraint it to specific versions, rather than have unlimited upper range. Right now the same would hold true for PHP 7.4, but it's not supported yet officially. I suspect we don't want to allow 7.4 yet.
So it would be better to list the version like in twig constraint that are supported:
`php: "7.1|^7.2|^7.3"`

mickaelandrieu commented:
> So it would be better to list the version like in twig constraint that are supported

watchout ! Twig 2 won't be supported anymore, for what it means.

I do agree that this would be a best practice, for projects that doesn't support current version of PHP

matks commented:
> The PrestaShop Core team has a very different point of view about "listing the supported versions" meaning. Currently, they don't consider that PrestaShop supports Twig 2 for instance, while it was listed in the previous Composer configuration (from 1.7.4 to 1.7.6).

Indeed, because PrestaShop is an application, not a library/dependency. So what actually matters is the composer.lock, not the composer.json.

This has already been discussed here:
https://github.com/PrestaShop/PrestaShop/pull/13761#issuecomment-522031661
https://github.com/PrestaShop/PrestaShop/pull/13761#issuecomment-526196326

We're sorry if a bad composer.json configuration introduced confusion (and we'll fix that), but the only thing we currently support is *the dependencies as listed in composer.lock*. They are the ones we run tests (manual and automated) with, they are the ones we work daily with, they are the ones we target, explore, monitor and control in order to assure the best experience for PrestaShop. Not the wide range of composer.json versions.

Progi1984 commented on classes/Mail.php:
```diff
@@ -347,22 +348,25 @@ function ($carry, $item) {
+                // mail() support was removed from SwiftMailer for security reasons
+                // previously => $connection = \Swift_MailTransport::newInstance();
+                // @see https://github.com/swiftmailer/swiftmailer/issues/866
+                $connection = false;
```
@PierreRambaud That sounds that mail() will not be supported with 1.7.7. May be to indicate somewhere.

PierreRambaud replied:
Woot, didn't check this part :scream: 
We must found a solution because of postfix server :/

PierreRambaud replied:
I add the alternative of Sendmail transport and fix the sendMailTest feature

Progi1984 reviewed (commented):
One feedback

jolelievre commented on src/PrestaShopBundle/Entity/AdminFilter.php:
```diff
@@ -83,7 +83,7 @@ class AdminFilter
      * @var string
      *
-     * @ORM\Column(name="filter_id", type="string", length=255)
+     * @ORM\Column(name="filter_id", type="string", length=191)
```
Why 191 specifically?

PierreRambaud replied:
Because of utf8mb4.
> InnoDB has a maximum index length of 767 bytes for tables that use COMPACT or REDUNDANT row format, so for utf8mb3 or utf8mb4 columns, you can index a maximum of 255 or 191 characters, respectively. If you currently have utf8mb3 columns with indexes longer than 191 characters, you must index a smaller number of characters.

jusurb replied:
or specify explicitly index length
`CREATE INDEX idx_filter_id ON admin_filter (filter_id(191));`
for most cases if data fits into index lookup it will be fast. But if in most cases data exceeds 191 indexed chars it will have to do more work and filter more rows after applying the index

PierreRambaud replied:
This is also why it's not needed to have a `varchar(255)` with an `index(191)`, otherwise the index shouldn't be on this column :thinking:

Robin-Fischer-PS commented:
Hi :)

Could you resolve the conflict on Mail.php ? Thanks !

mvorisek commented:
New minor version of Doctrine ORM v2.7 has been released. Please upgrade the composer.lock `doctrine/orm` to `2.7.0`. The new version is like `2.6.x` with some deprecations and if the Doctrine version is bumped in this PR, it should target for the latest stable release.

Commit:
Update doctrine/orm to 2.7

PierreRambaud commented:
> New minor version of Doctrine ORM v2.7 has been released. Please upgrade the composer.lock `doctrine/orm` to `2.7.0`. The new version is like `2.6.x` with some deprecations and if the Doctrine version is bumped in this PR, it should target for the latest stable release.

Thanks for the suggestion :wink:

mvorisek commented:
Please fix also this line:
https://github.com/PrestaShop/PrestaShop/blob/4ca6874eefed0925278e435ec44df26d9a117674/classes/ImageType.php#L200
```
-        if (strstr($name, $theme_name) && self::getByNameNType($name)) {
+        if ($theme_name !== null && strstr($name, $theme_name) && self::getByNameNType($name)) {
```
The `strstr` otherwise emits a deprecation warning as of PHP 7.3 if the needle is not a string (`null` here if the `$theme_name` /  `Context::getContext()->shop->theme_name` is `null`)

Commit:
Remove deprecated warning when  is null

PierreRambaud commented:
@mvorisek Thanks, it's done :)

mvorisek commented:
As PHP 7.4 is out and this PR is still not merged, can you test to add PHP 7.4 to the tests? If there will be only several changes needed, I think it is good idea to cover the PHP 7.4 in this PR too to make PrestaShop compatible with the latest PHP in the upcoming Prestashop `1.7.7` release.

eternoendless commented:
PHP 7.4 compatibility is unfortunately out of scope for 1.7.7, but is due for 1.7.8. That doesn't necessarily mean that it _won't_ work in 1.7.7, it just means that we won't _ensure_ that it works in this version.

Commit:
Update phpunit tests and docblocks

Commit:
Cast value to float in credit slip tests

jolelievre commented:
Thank you @mickaelandrieu and @PierreRambaud
